tiatoolbox/models/architecture/micronet.py

```diff
@@ -0,0 +1,567 @@
+
+            """
+            module_dict = OrderedDict()
+            module_dict["up1"] = nn.ConvTranspose2d(
+                in_ch, out_ch, kernel_size=up_kernel, stride=up_strides
+            )
+            module_dict["conv1"] = nn.Sequential(
+                nn.Conv2d(
+                    out_ch,
+                    out_ch,
+                    kernel_size=(3, 3),
+                    stride=(1, 1),
+                    padding=0,
+                    bias=True,
+                ),
+                nn.Tanh(),
+            )
+            return nn.ModuleDict(module_dict)
+
+        def out_branch(in_ch):
+            """Group5 branch for MicroNet.
+
+            Args:
+                in_ch (int): Number of input channels.
+
+            Returns:
+                torch.nn.Sequential: An output of type :class:`torch.nn.Sequential`
+
+            """
+            return nn.Sequential(
+                nn.Dropout2d(p=0.5),
+                nn.Conv2d(
+                    in_ch,
+                    self.__num_class,
+                    kernel_size=(3, 3),
+                    stride=(1, 1),
+                    padding=0,
+                    bias=True,
+                ),
+                nn.Softmax(),
+            )
+
+        module_dict = OrderedDict()
+        module_dict["b1"] = group1_branch(num_input_channels, num_input_channels, 64)
+        module_dict["b2"] = group1_branch(128, num_input_channels, 128)
+        module_dict["b3"] = group1_branch(256, num_input_channels, 256)
+        module_dict["b4"] = group1_branch(512, num_input_channels, 512)
+
+        module_dict["b5"] = group2_branch(1024, 2048)
+
+        module_dict["b6"] = group3_branch(2048, 1024, 1024)
+        module_dict["b7"] = group3_branch(1024, 512, 512)
+        module_dict["b8"] = group3_branch(512, 256, 256)
+        module_dict["b9"] = group3_branch(256, 128, 128)
+
+        module_dict["fm1"] = group4_branch(128, 64, (2, 2), (2, 2))
+        module_dict["fm2"] = group4_branch(256, 128, (4, 4), (4, 4))
+        module_dict["fm3"] = group4_branch(512, 256, (8, 8), (8, 8))
+
+        module_dict["aux_out1"] = out_branch(64)
+        module_dict["aux_out2"] = out_branch(128)
+        module_dict["aux_out3"] = out_branch(256)
+
+        module_dict["out"] = out_branch(64 + 128 + 256)
+
+        self.layer = nn.ModuleDict(module_dict)
+
+        self.apply(weights_init)
+
+    def forward(self, input_tensor: torch.Tensor):  # skipcq: PYL-W0221
+        """Logic for using layers defined in init.
+
+        This method defines how layers are used in forward operation.
+
+        Args:
+            input_tensor (torch.Tensor): Input images, the tensor is
+                in the shape of NCHW.
+
+        Returns:
+            list: A list of main and auxiliary outputs.
+                The expected format is [main_output, aux1, aux2, aux3].
+
+        """
+
+        def group1_branch(layer, in_tensor, resized_feat):
```

Comment:
Would the below group functions be better defined in __init__ under a slight different name such as group_link? Makes forward quite clunky otherwise.

Reply:
Addressed in this commit https://github.com/TissueImageAnalytics/tiatoolbox/pull/293/commits/9af1c008012edf59b1599184b900515b412365d0